src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala

```diff
@@ -4,109 +4,137 @@ import vsys.blockchain.transaction.ValidationError
       case _ => Left(ContractUnsupportedOPC)
     } else Left(ContractDataTypeMismatch)
 
-  def sqrt(x: DataEntry, dataStack: Seq[DataEntry], pointer: Byte): Either[ValidationError, Seq[DataEntry]] =
-    if (pointer > dataStack.length || pointer < 0) Left(ContractLocalVariableIndexOutOfRange)
-    else x.dataType match {
+  def sqrt(x: DataEntry): Either[ValidationError, DataEntry] =
+    x.dataType match {
       case BigInteger => {
         val value = BigInteger.deserializer(x.data)
         if (value < 0) Left(ValidationError.OverflowError)
-        else appendRes[BigInt](pointer, sqrtBigInt(value),  BigInteger, dataStack)
+        else formatRes[BigInt](sqrtBigInt(value), BigInteger)
       }
       case _ => Left(ContractUnsupportedOPC)
     }
 
-  // babylonian method
-  def sqrtBigInt(y: BigInt): BigInt = {
-    if (y > 3) Stream.iterate((y, (y >> 1) + 1)){ case (z, x) => (x, (y / x + x) >> 1) }.dropWhile{ case(z, x) => x < z }.head._1
-    else if (y > 0) 1 else 0
-  }
-  def concat(x: DataEntry, y: DataEntry, dataStack: Seq[DataEntry], pointer: Byte): Either[ValidationError, Seq[DataEntry]] = {
-    if (pointer > dataStack.length || pointer < 0) {
-      Left(ContractLocalVariableIndexOutOfRange)
-    } else {
-      for {
-        res <- DataEntry.create(x.data ++ y.data, DataType.ShortBytes)
-      } yield dataStack.patch(pointer, Seq(res), 1)
+  def boolBiOperation(x: DataEntry, y: DataEntry, operator: BoolByteBiOperator): Either[ValidationError, DataEntry] =
+    if (x.dataType == y.dataType && x.dataType == Boolean) formatResB[Boolean](Array(operator.op(x, y)), Boolean)
+    else Left(ContractInvalidOPCData)
+
+  def not(x: DataEntry): Either[ValidationError, DataEntry] =
+    x.dataType match {
+      case Boolean => formatResB[Boolean](Array(bNot(x)), Boolean)
+      case _ => Left(ContractUnsupportedOPC)
     }
-  }
 
-  def constantGet(constant: Array[Byte], dataStack: Seq[DataEntry], pointer: Byte): Either[ValidationError, Seq[DataEntry]] = {
-    if (pointer > dataStack.length || pointer < 0) {
-      Left(ContractLocalVariableIndexOutOfRange)
-    } else {
-      DataEntry.fromBytes(constant) match {
-        case Right(v) => Right(dataStack.patch(pointer, Seq(v), 1))
-        case Left(e) => Left(e)
+  def convertion(x: DataEntry, t: DataEntry): Either[ValidationError, DataEntry] =
+    t.dataType match {
+      case DataTypeObj => x.dataType match {
+        case Int32 | Amount | Timestamp | BigInteger => {
+          val to: DataTypeVal[_] = DataTypeObj.deserializer(t.data)
+          to match {
+            case Int32      => formatResB[Int]   (x.data, Int32)
+            case Amount     => formatResB[Long]  (x.data, Amount)
+            case Timestamp  => formatResB[Long]  (x.data, Timestamp)
+            case BigInteger => formatResB[BigInt](x.data, BigInteger)
+            case _ => Left(ContractUnsupportedOPC)
+          }
+        }
+        case _ => Left(ContractUnsupportedOPC)
```

Comment:
A little bit confused about the usage of `ContractUnsupportedOPC` and `ContractInvalidOPCData ` here.
case 1: no such opc index => `ContractUnsupportedOPC`
case 2: have such opc index but not support the data type => ? ContractUnsupportedOPC or ContractInvalidOPCData
case 3: have such opc index, but have relationship between the input dataType => ContractDataTypeMismatch

Reply:
so case 2 should return `ContractUnsupportedOPC`, and
should clarify the usage of `ContractInvalidOPCData `

Reply:
Yes, it's case 2 both L87 and L84, if converting FROM other datatypes will be supported in future, it is `ContractUnsupportedOPC`.  it's also case 2 L84, if converting TO other datatypes will be supported in future, it is also `ContractUnsupportedOPC`

Reply:
I see, agree to use the `ContractInvalidOPCData` in L89. And it is also a single usage here. L66 should be updated.

Reply:
In my mind, case 2 was be separated to two sub cases:
1. reasonable and may be supported in future: `ContractUnsupportedOPC`
2. unreasonable datatype: `ContractInvalidOPCData`

Maybe there is mistake in my understanding....

Reply:
You're right. I think it clarifies some boundary definitions in my mind.